### ISSUE TITLE
Why the implementation of SWA is different with the original paper?

### ISSUE THREAD

**Issue by khanrc** [gcp/leela-zero#1264]
The original paper did not use the (exponential) moving average, but the just average (arithmetic mean). But SWA in LZ is implemented by moving average (exactly speaking, LZ has limitation of `swa_max_n` which makes average to moving average). 

Is there a specific reason for doing this? I tried to search but couldn't find the reason.

**Comment by gcp** [comment]
Doesn't the paper say SWA should only be applied after the network has done it's initial convergence? That's what the implementation achieves: it only considers the last networks.

**Comment by khanrc** [comment]
Right. But the EMA (exponential moving average) has a bias to the most recent weight, so I think just average like original paper could be better. What do you think about this?

**Comment by Ishinoshita** [comment]
If the main concept of SWA is based on the observation that the network, in  the weight space, is boucing around the side of a bassin, unable to hit the minimum, then there is no reason to favour last iteration other previous ones, and arithmetic average makes more sense. In the case of RL-generated samples, i.e. generated by ever changing player, it might be that the bassin shape / bassin minimum are changing/ moving with iterations, in which case EMA might make more sense. Intuitively, if MA has some momentum, them EMA might do better.

**Comment by gcp** [comment]
It seems this was changed as part of cfed4e47d3b37719742f99aaec776730b358ea58, the initial implementation did use normal averages. @ywrt 

>In the case of RL-generated samples, i.e. generated by ever changing player

The training does reset after every dump of training data, so I think that consideration does not apply here, i.e. after an initial convergence period, the state should be static.

**Comment by Ishinoshita** [comment]
@gcp  Thank you for this explanation of how SWA is implemented in LZ.

**Comment by Tsun-hung** [comment]
@gcp  I notice that the networks of 8K and 16K steps are in bad performance, so is it worth considering to stop putting these two networks on the match?


![49](https://user-images.githubusercontent.com/33803485/39301965-a79c8b92-4983-11e8-8e14-876acef1e99e.jpg)
![45](https://user-images.githubusercontent.com/33803485/39301968-ac0249ba-4983-11e8-97e7-cb6e5c49010e.jpg)
![42](https://user-images.githubusercontent.com/33803485/39301974-afa9034c-4983-11e8-9fcd-22388fd61873.jpg)
![21](https://user-images.githubusercontent.com/33803485/39301979-b4ba37de-4983-11e8-9484-44c3770aae10.png)
![13](https://user-images.githubusercontent.com/33803485/39301986-b87e8384-4983-11e8-84bd-7eb3a236f9e6.jpg)
![05](https://user-images.githubusercontent.com/33803485/39301990-bc14e254-4983-11e8-9c05-3690b556c948.jpg)

**Comment by gcp** [comment]
They don't use SWA, so that's totally off-topic here.

**Comment by Tsun-hung** [comment]
Sorry, I notice the thread is about network and just too lazy to open a new issue

> They don't use SWA, so that's totally off-topic here.

**Comment by ywrt** [comment]
The implementation is the same before and after cfed4e4: It uses a uniformly weighted arithmetic mean until self.swa_max_n samples have been reached, and there-after it uses an exponentially weighted moving average (EMA) with tau=self.swa_max_n

The SWA assumption is that, after convergence, training is sampling from a Gaussian centered at the optimum.

With this assumption, using an EMA with tau=N is the same as using an arithmetic mean over ~0.73*N samples. (As measured by the L2 distance to the center of the Gaussian)

However! The SWA assumption is a brave, bold and courageous assumption, and even so, it only applied after convergence, so it may be in reality there is a difference between an EWMA and a uniformly weighted average.

**Comment by khanrc** [comment]
@Ttl @ywrt Oh, I understand! I missed the fact that training is end at 128k step, thanks. 

My last concern is that the step 8k, which is the first checkpoint step, is initial convergence? I think it could be too early to converge. 

And, besides this topic, why is training restarted at 128k step? More training does not help?

**Comment by Ishinoshita** [comment]
Is there a way to identify which networks were SWA-originated, in the "Best Network Hash" list ?

**Comment by gcp** [comment]
>And, besides this topic, why is training restarted at 128k step? More training does not help?

At some point the network overfits and more training makes it worse. But SWA seems very effective at preventing this, and the best networks all have been at 128k steps. So I increased the max steps count yesterday (but effects won't be visible until later today).

**Comment by ssj-gz** [comment]
@gcp Will that be 192K and 256k like before?

**Comment by gcp** [comment]
It's not *particularly* interesting, especially as you're using a lower learning rate than the training machine, so a slightly better fit wouldn't be unexpected, I guess. We've also had previous reports that narrowing the training window can give a good boost, but it also risks over-fitting, see issues with previous bootstraps etc. It's just too little "data" to draw conclusions from.

>@gcp Will that be 192K and 256k like before?

I'll just try up to 192k first.

>Is there a way to identify which networks were SWA-originated, in the "Best Network Hash" list ?

If you download them it's indicated in the name.

**Comment by MaxMaki** [comment]
Ok! What's the current training rate in the training machine?

Anyways, i'll keep training just for my own curiosity, and in case it somehow as a miracle ends up a great deal stronger in my own testing i'll post.

**Comment by gcp** [comment]
0.00035 @ bs = 256

**Comment by MaxMaki** [comment]
Ok, thanks for the info!

**Comment by ywrt** [comment]
FWIW, I think that sma_max_n == 16 is far too low. A value of around 2000 seems a reasonable comprise between smoothing and numerical stability.

(This is based the empirical observation that the ratio between the standard deviation and the mean in the gradient is larger than 100x by step ~200k)

**Comment by Ttl** [comment]
In the constant learning rate experiments in the SWA paper it took only about 10 averages before the test error of the averaged network plateaued. Going beyond 20 averages did not change the test error.

**Comment by ssj-gz** [comment]
192k looking decent :)

**Comment by ssj-gz** [comment]
@gcp The last couple of 192k-steppers have been close - time to raise the ceiling a little? :)

**Comment by gcp** [comment]
Has already been done, but such changes always take a full cycle to become active.

**Comment by afalturki** [comment]
@khanrc I think training always restarts from the current best network.

**Comment by Mardak** [comment]
Looks like the newly added +256k was successful (along with new submit-match, verification, recent graph, and test networks!):

<img width="1031" alt="screen shot 2018-05-01 at 8 45 25 pm" src="https://user-images.githubusercontent.com/438537/39504507-eb7cc880-4d80-11e8-86a6-4d7c78e30864.png">

**Comment by gcp** [comment]
I think the more steps SWA has to average, the better it works. It's not conclusive to me right now whether taking an SWA average more often is better than running more steps (of course the former goes faster!).

I suspect that we're getting best results with those high step SWA networks partially because the learning rate is fairly high. But as you know, I prefer it that way to avoid overfits as much as possible, and SWA gives you the best of both worlds as long as you are willing to increase max steps. I think 256k is probably the maximum though, else the lag between incorporating new games in the training is too high. 

I'm not sure I understood the second question, there is no real interaction between the training steps and the size of the weight file.

**Comment by jkiliani** [comment]
Testing the effect of the SWA sampling interval sounds like a rather simple experiment... for that you could simply reduce the sampling interval tfprocess.py saves weight files at, and then compare the SWA nets at the same step count, between the original sampling interval and the reduced one.

Barring implementation bugs and disk space limitations, I don't see what could go wrong with that.

**Comment by marcocalignano** [comment]
Another experiment would be to Average the first trained network (8K) with the actual best network. And another to always average the best network in the SWA. (if @gcp runs out of experiment to do! :smile: )
